From 5092aec1e16abfb0a2626df11f5fda8b44e6f184 Mon Sep 17 00:00:00 2001
From: Li Wencheng <liwencheng@phytium.com.cn>
Date: Wed, 7 May 2025 17:22:12 +0800
Subject: [PATCH 42/67] PHYTIUM: net/phytmac: Bugfix set WOL failed issue

Before configuring WOL, we need to obtain the packet types
that WOL supports.

Signed-off-by: Li Wencheng <liwencheng@phytium.com.cn>
Signed-off-by: Wang Yinfeng <wangyinfeng@phytium.com.cn>
Change-Id: I22ed39f691044f35fa09ee76788711fc723fc6fc

Link: https://gitee.com/phytium_embedded/phytium-linux-kernel/commit/b44b37fdc57a32edb8bf0a5e763ea71c716cce09
Signed-off-by: Kexy Biscuit <kexybiscuit@aosc.io>
---
 .../net/ethernet/phytium/phytmac_ethtool.c    | 20 ++++++++-----------
 1 file changed, 8 insertions(+), 12 deletions(-)

diff --git a/drivers/net/ethernet/phytium/phytmac_ethtool.c b/drivers/net/ethernet/phytium/phytmac_ethtool.c
index 2c0bbb8f2772..bac59327087f 100644
--- a/drivers/net/ethernet/phytium/phytmac_ethtool.c
+++ b/drivers/net/ethernet/phytium/phytmac_ethtool.c
@@ -90,24 +90,20 @@ static void phytmac_get_wol(struct net_device *ndev, struct ethtool_wolinfo *wol
 {
 	struct phytmac *pdata = netdev_priv(ndev);

+	wol->wolopts = 0;
 	phylink_ethtool_get_wol(pdata->phylink, wol);

-	if (pdata->wol & PHYTMAC_WAKE_MAGIC) {
+	wol->supported = WAKE_MAGIC | WAKE_ARP |
+			 WAKE_UCAST | WAKE_MCAST;
+
+	if (pdata->wol & PHYTMAC_WAKE_MAGIC)
 		wol->wolopts |= WAKE_MAGIC;
-		wol->supported |= WAKE_MAGIC;
-	}
-	if (pdata->wol & PHYTMAC_WAKE_ARP) {
+	if (pdata->wol & PHYTMAC_WAKE_ARP)
 		wol->wolopts |= WAKE_ARP;
-		wol->supported |= WAKE_ARP;
-	}
-	if (pdata->wol & PHYTMAC_WAKE_UCAST) {
+	if (pdata->wol & PHYTMAC_WAKE_UCAST)
 		wol->wolopts |= WAKE_UCAST;
-		wol->supported |= WAKE_UCAST;
-	}
-	if (pdata->wol & PHYTMAC_WAKE_MCAST) {
+	if (pdata->wol & PHYTMAC_WAKE_MCAST)
 		wol->wolopts |= WAKE_MCAST;
-		wol->supported |= WAKE_MCAST;
-	}
 }

 static int phytmac_set_wol(struct net_device *ndev, struct ethtool_wolinfo *wol)
--
2.50.1
